[Split CSSPropertyParserHelpers] Part 2: String, URL, Position and Image

weinig opened the pull request:
#### 24e2dca84e285abddbd5870d9d8e3ea4bf0ff2b6
<pre>
[Split CSSPropertyParserHelpers] Part 2: String, URL, Position and Image
<a href="https://bugs.webkit.org/show_bug.cgi?id=275780">https://bugs.webkit.org/show_bug.cgi?id=275780</a>

Reviewed by Darin Adler.

Further split up CSSPropertyParserHelpers into type specific files.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.h:
    - Move custom and dashed ident consumers.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+String.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+String.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
    - Move consume functions to new files.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
    - Remove moved functions.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
* Source/WebCore/css/process-css-properties.py:
    - Update includes.

Canonical link: <a href="https://commits.webkit.org/280294@main">https://commits.webkit.org/280294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92f4427d592c1bdb7729c26d262ba9dfe5c24b4a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56200 "28 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59806 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6636 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43148 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6830 "Hash 92f4427d for PR 30086 does not build (failure)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45445 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4610 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58229 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48483 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26374 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30196 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5813 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5640 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52197 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61489 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/108 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/6830 "Hash 92f4427d for PR 30086 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52762 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/108 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48550 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52600 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12451 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/96 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31353 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32439 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.cpp:
```diff
@@ -59,5 +60,37 @@ RefPtr<CSSPrimitiveValue> consumeIdentRange(CSSParserTokenRange& range, CSSValue
+{
+    auto rangeCopy = range;
+    auto result = consumeCustomIdent(range, shouldLowercase);
+    if (result && result->stringValue().startsWith("--"_s))
+        return result;
+    range = rangeCopy;
+    return nullptr;
```
I probably would have reversed this so that the error case is the early return, but that would add braces and maybe even predict the branch wrong. But still, that’s what comes to mind

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+
+// MARK: Generate Image
+
+static bool isGeneratedImage(CSSValueID id)
```
What helps us keep it so that this function has all the values that it should, and none that it shouldn’t? Is there a test that can check this?

weinig replied:
It's kind of silly to have it at all, since the next line just switches over those again. Will remove.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+        if ((horizontal && consumeIdent<CSSValueRight>(range)) || (!horizontal && consumeIdent<CSSValueBottom>(range)))
+            return CSSPrimitiveValue::create(100., CSSUnitType::CSS_PERCENTAGE);
+        if (consumeIdent<CSSValueCenter>(range))
+            return CSSPrimitiveValue::create(50., CSSUnitType::CSS_PERCENTAGE);
```
Do you need these decimal points? Can’t you just pass an integer and have it be automatically converted to a `double`?

weinig replied:
No need for that. Just copy/paste. Fixed.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+
+static std::optional<CSSGradientDeprecatedPoint> consumeDeprecatedGradientPoint(CSSParserTokenRange& range)
+{
+    auto x = consumeDeprecatedGradientPointValue(range, true);
```
This is that bad pattern where we use a mystery boolean, and we typically try to use enumerations or something else instead. Presumably just moved code, though.

weinig replied:
Yeah copied, but fixed now.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+    if (id != CSSValueFrom && id != CSSValueTo && id != CSSValueColorStop)
+        return false;
+
+    CSSParserTokenRange args = consumeFunction(range);
```
Maybe `auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+
+static bool consumeDeprecatedGradientColorStop(CSSParserTokenRange& range, CSSGradientColorStop& stop, const CSSParserContext& context)
+{
+    CSSValueID id = range.peek().functionId();
```
Maybe `auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+{
+    CSSValueID id = range.peek().functionId();
+    if (id != CSSValueFrom && id != CSSValueTo && id != CSSValueColorStop)
+        return false;
```
Maybe `switch`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+        }
+    }
+
+    CSSPrefixedRadialGradientValue::GradientBox gradientBox = std::monostate { };
```
Is `= std::monostate { }` needed? Why isn’t that what the default constructor makes?

weinig replied:
I prefer the explicit construction since the default value of a variant is based on position and an innocuous looking change to the order in the definition would potentially change behavior.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+    if (!stops)
+        return nullptr;
+
+    auto computedColorInterpolationMethod = computeGradientColorInterpolationMethod(colorInterpolationMethod, *stops);
```
Not sure we need the local variable for this.

weinig replied:
I think we do. `stops` is WTFMove()'d in the function call, and since the order of evaluation of function call arguments is not defined, stops might be moved before this would have been called.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+            if (position)
+                data.gradientBox = CSSRadialGradientValue::GradientBox { std::in_place_type<CSSGradientPosition>, WTFMove(*position) };
+            else
+                data.gradientBox = std::monostate { };
```
If we just wrote `= { }`, would this work?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+            return nullptr;
+    }
+
+    CSSConicGradientValue::Angle angle = std::monostate { };
```
Is `= std::monostate { }` needed? Why isn’t that what the default constructor makes?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+    if (!stops)
+        return nullptr;
+
+    auto computedColorInterpolationMethod = computeGradientColorInterpolationMethod(colorInterpolationMethod, *stops);
```
Not sure we need the local variable for this.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+    if (!stops)
+        return nullptr;
+
+    auto computedColorInterpolationMethod = computeGradientColorInterpolationMethod(colorInterpolationMethod, *stops);
```
Not sure we need the local variable for this.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+    auto filterValue = consumeFilter(args, context, AllowedFilterFunctions::PixelFilters);
+
+    if (!filterValue)
+        return nullptr;
```
I’d omit the blank line here.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+    else if (id == CSSValueWebkitFilter || id == CSSValueFilter)
+        result = consumeFilterImage(args, context);
+    else if (id == CSSValuePaint)
+        result = consumeCustomPaint(args, context);
```
`switch`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -0,0 +1,1185 @@
+    }
+
+    if (range.peek().type() == FunctionToken) {
+        CSSValueID functionId = range.peek().functionId();
```
`auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.h:
```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <wtf/OptionSet.h>
+#include <wtf/RefPtr.h>
```
Shouldn't need this include just to declare functions that return a `RefPtr`. `Forward.h` should be sufficient.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
```diff
@@ -0,0 +1,279 @@
+            resultY = WTFMove(center);
+    }
+
+    ASSERT(resultX && resultY);
```
I’d omit this. The `releaseNonNull` includes these kinds of assertions.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
```diff
@@ -0,0 +1,279 @@
+        if (!currentValue->isValueID())
+            return std::nullopt;
+
+        CSSValueID id = currentValue->valueID();
```
`auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h:
```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include "CSSPropertyParserConsumer+Primitives.h"
+#include <wtf/Forward.h>
```
Don’t need this because `Ref.h` includes it.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h:
```diff
@@ -0,0 +1,62 @@
+
+#include "CSSPropertyParserConsumer+Primitives.h"
+#include <wtf/Forward.h>
+#include <wtf/Ref.h>
```
Don’t need this because `CSSPropertyParserConsumer+Primitives.h` includes it.

weinig replied:
After doing a bunch of refactoring, I am a bit hesitant to rely on knowledge like this, as, especially with the unified file builds, it makes some small changes break things in unpredictable ways.

That said, it is pretty established, so I made the change, but it has me thinking.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h:
```diff
@@ -0,0 +1,62 @@
+
+namespace WebCore {
+
+class CSSParserTokenRange;
```
Don’t need this because `CSSPropertyParserConsumer+Primitives.h` forward declares it.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+String.h:
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <wtf/Forward.h>
+#include <wtf/RefPtr.h>
```
Should not need this include to declare a function that returns a `RefPtr`.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    if (token.functionId() == CSSValueUrl) {
+        CSSParserTokenRange urlRange = range;
```
`auto`?

weinig replied:
This is one of those sharp edges of c++ that after coming back to I always have to lookup. Specifically:

```
MyClass& variable = ...;
auto foo = variable;
```

does `auto` resolve to `MyClass` or `MyClass&`?

In this case, we really want it to be `MyClass` as the intent is to copy. I am pretty sure it does, but its definitely one of those places where implicitness gets me.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
```diff
@@ -0,0 +1,71 @@
+
+StringView consumeURLRaw(CSSParserTokenRange& range)
+{
+    const CSSParserToken& token = range.peek();
```
`auto&`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
```diff
@@ -0,0 +1,71 @@
+
+    if (token.functionId() == CSSValueUrl) {
+        CSSParserTokenRange urlRange = range;
+        CSSParserTokenRange urlArgs = urlRange.consumeBlock();
```
`auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
```diff
@@ -0,0 +1,71 @@
+    if (token.functionId() == CSSValueUrl) {
+        CSSParserTokenRange urlRange = range;
+        CSSParserTokenRange urlArgs = urlRange.consumeBlock();
+        const CSSParserToken& next = urlArgs.consumeIncludingWhitespace();
```
`auto&`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
```diff
@@ -0,0 +1,71 @@
+
+RefPtr<CSSPrimitiveValue> consumeURL(CSSParserTokenRange& range)
+{
+    StringView url = consumeURLRaw(range);
```
`auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.h:
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <wtf/Forward.h>
+#include <wtf/RefPtr.h>
```
Should not need this include to declare a function that returns a `RefPtr`.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -126,1402 +117,54 @@ namespace CSSPropertyParserHelpers {
-        .unitlessZero = UnitlessZeroQuirk::Allow
+#if !ENABLE(VARIATION_FONTS)
+    auto isIntegerAndDivisibleBy100 = [](double value) {
+        ASSERT(value >= 1 && value <= 1000);
```
I'd break this into two separate assertions, not sure other people agree, though.

darinadler reviewed (approved):
As usual, I reviewed all the moved code, because I can’t really tell if it was just moved or changed in the process.

weinig committed:
[Split CSSPropertyParserHelpers] Part 2: String, URL, Position and Image
https://bugs.webkit.org/show_bug.cgi?id=275780

Reviewed by Darin Adler.

Further split up CSSPropertyParserHelpers into type specific files.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.h:
    - Move custom and dashed ident consumers.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+String.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+String.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.h: Copied from Source/WebCore/css/parser/CSSPropertyParserHelpers.h.
    - Move consume functions to new files.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
    - Remove moved functions.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
* Source/WebCore/css/process-css-properties.py:
    - Update includes.

Canonical link: https://commits.webkit.org/280294@main
